Gravatar/GravatarService.cs

```diff
@@ -44,111 +50,69 @@ public GravatarService(IImageCache imageCache)
         {
         }
 
-        /// <summary>
-        /// Loads avatar either from the local cache or from the remote service.
-        /// </summary>
-        public async Task<Image> GetAvatarAsync(string email, int imageSize, string defaultImageType)
+        async Task<Image> IAvatarService.GetAvatarAsync(string email, int imageSize, string defaultImageType)
         {
+            if (email == null)
+            {
+                throw new ArgumentNullException(nameof(email));
+            }
+
+            if (defaultImageType == null)
+            {
+                throw new ArgumentNullException(nameof(defaultImageType));
+            }
+
             var imageFileName = _avatarImageNameProvider.Get(email);
-            var image = await _cache.GetImageAsync(imageFileName, null);
+
+            var image = await _cache.GetImageAsync(imageFileName);
+
             if (image == null)
             {
                 if (email.EndsWith(GitHubPrivateEmailSuffix, StringComparison.OrdinalIgnoreCase))
                 {
-                    image = await LoadFromGitHubAsync(imageFileName, email, imageSize);
+                    image = await LoadFromGitHubAsync(email, imageSize);
                 }
                 else
                 {
-                    image = await LoadFromGravatarAsync(imageFileName, email, imageSize, GetDefaultImageType(defaultImageType));
+                    image = await LoadFromGravatarAsync(email, imageSize, ((IAvatarService)this).GetDefaultImageType(defaultImageType));
```

Comment:
> `((IAvatarService)this)`

This is why I prefer `<inheritdoc />` to what's in use now. Explicit interface implementation tells me one of the following:

1. The interface is an `internal` interface, the method exposes `internal` types in the signature, but the containing type is `public`
2. The method duplicates another method in the type with similar but not equal signatures, e.g. `IEnumerable<T>.GetEnumerator` and `IEnumerable.GetEnumerator`
3. The method is part of a legacy interface, and implementing it requires use of another legacy method, but it's not desirable to expose them to callers

Reply:
In GE there is a lot of use of concrete classes, rather than the interfaces they implement. The interfaces don't cover all the public members, and many of their members could be removed due to non-use through the interface.

Making the implementation explicit means you _have_ to use the interface to call the member. Compare the number of members in `IGitModule` vs `GitModule`.

Reply:
I don't see the advantage here. The only purpose I see it serving in this type is making the implementation more clumsy.

Reply:
:memo: Also note that there are some refactorings which can cause this approach to hide an `InvalidCastException` that the compiler would otherwise be able to alert developers to. Consider the different between the form used in the PR and this:

```csharp
IAvatarService avatarService = this; // fails to compile if interface is wrong
```